### PR TITLE
Fix UniqueContraintViolation and symbol (:gre, :mac2mac) mysql errors

### PR DESCRIPTION
**Pull request by Mopster**
Prevents the following error in the vnmgr,log

```
I, [2015-04-23T06:03:30.392090 #10642]  INFO -- : failed to execute: class_name: tunnel method_name: execute args: [:create, {:src_datapath_id=>1, :dst_datapath_id=>2, :src_interface_id=>1, :dst_interface_id=>2, :mode=>:mac2mac}]
I, [2015-04-23T06:03:30.392181 #10642]  INFO -- : Mysql2::Error: Duplicate entry '1-2-1-2-0' for key 'tunnels_datapath_id_interface_id_index' (Sequel::UniqueConstraintViolation)
```
which resulted in the following error in the vna.log 

```
E, [2015-04-23T06:03:30.758691 #4820] ERROR -- : couldn't decode message
DCell::Server::InvalidMessageError: invalid message: undefined class/module Mysql2::
	/opt/axsh/openvnet/vnet/vendor/bundle/ruby/2.1.0/gems/dcell-0.15.0/lib/dcell/server.rb:59:in `rescue in decode_message'
	/opt/axsh/openvnet/vnet/vendor/bundle/ruby/2.1.0/gems/dcell-0.15.0/lib/dcell/server.rb:56:in `decode_message'
	/opt/axsh/openvnet/vnet/vendor/bundle/ruby/2.1.0/gems/dcell-0.15.0/lib/dcell/server.rb:38:in `handle_message'
	/opt/axsh/openvnet/vnet/vendor/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/calls.rb:25:in `public_send'
	/opt/axsh/openvnet/vnet/vendor/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/calls.rb:25:in `dispatch'
	/opt/axsh/openvnet/vnet/vendor/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/calls.rb:122:in `dispatch'
	/opt/axsh/openvnet/vnet/vendor/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/actor.rb:322:in `block in handle_message'
	/opt/axsh/openvnet/vnet/vendor/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/actor.rb:416:in `block in task'
	/opt/axsh/openvnet/vnet/vendor/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/tasks.rb:55:in `block in initialize'
	/opt/axsh/openvnet/vnet/vendor/bundle/ruby/2.1.0/gems/celluloid-0.15.2/lib/celluloid/tasks/task_fiber.rb:13:in `block in create'
```

### CODE REVIEW

File: vnet/lib/vnet/core/tunnel_manager.rb [hunk 1]
```diff
@@ -295,7 +297,17 @@ def create_tunnel(options, tunnel_mode)
       info log_format("creating tunnel entry mode '#{tunnel_mode}'",
                       options.map { |k, v| "#{k}:#{v}" }.join(" "))
 
-      tunnel = MW::Tunnel.create(options.merge(mode: tunnel_mode))
+      if @mutex_create_tunnel.locked?
```
Comment (hansode) [inline vnet/lib/vnet/core/tunnel_manager.rb]: [reek] [DuplicateMethodCall] calls @mutex_create_tunnel.locked? 2 times

File: vnet/lib/vnet/core/tunnel_manager.rb [hunk 2]
```diff
@@ -295,7 +297,17 @@ def create_tunnel(options, tunnel_mode)
...
+        while @mutex_create_tunnel.locked? | @mutex_create_tunnel.owned? do
```
Comment (hansode) [inline vnet/lib/vnet/core/tunnel_manager.rb]: [reek] [DuplicateMethodCall] calls @mutex_create_tunnel.locked? 2 times

**Comment by akry** [comment]
What about ignoring the error? Only if the query is of tunnel, it is anyway successfully created.

**Comment by Mopster** [comment]
I thought of ignoring the error, but an error is an error, it should never really be ignored unless there is a very good reason. The cause of the error can be different so I don't want to ignore all the different causes.

File: vnet/lib/vnet/node_api/tunnel.rb [hunk 3]
```diff
@@ -15,17 +16,38 @@ def update_mode(id, mode)
...
+        tunnel = super({
+          :src_datapath_id => options[:src_datapath_id],
+          :dst_datapath_id => options[:dst_datapath_id],
+          :src_interface_id => options[:src_interface_id],
+          :dst_interface_id => options[:dst_interface_id]
+        })
+      end
+
+      def create_or_find(options)
+        # Only create a tunnel if it doesn't exist yet
```
Comment (hansode) [inline vnet/lib/vnet/node_api/tunnel.rb]: [rails_best_practices] remove trailing whitespace

File: vnet/lib/vnet/core/tunnel_manager.rb [hunk 4]
```diff
@@ -105,14 +107,14 @@ def query_filter_from_params(params)
     def select_tunnel_mode(src_interface_id, dst_interface_id)
       src_interface = @interfaces[src_interface_id]
       dst_interface = @interfaces[dst_interface_id]
-      return :unknown unless src_interface && src_interface[:network_id]
-      return :unknown unless dst_interface && dst_interface[:network_id]
+      return "unknown" unless src_interface && src_interface[:network_id]
+      return "unknown" unless dst_interface && dst_interface[:network_id]
```
Comment (hansode) [inline vnet/lib/vnet/core/tunnel_manager.rb]: [reek] [DuplicateMethodCall] calls dst_interface[:network_id] 3 times

File: vnet/lib/vnet/core/tunnel_manager.rb [hunk 5]
```diff
@@ -105,14 +107,14 @@ def query_filter_from_params(params)
     def select_tunnel_mode(src_interface_id, dst_interface_id)
       src_interface = @interfaces[src_interface_id]
       dst_interface = @interfaces[dst_interface_id]
-      return :unknown unless src_interface && src_interface[:network_id]
-      return :unknown unless dst_interface && dst_interface[:network_id]
+      return "unknown" unless src_interface && src_interface[:network_id]
```
Comment (hansode) [inline vnet/lib/vnet/core/tunnel_manager.rb]: [reek] [DuplicateMethodCall] calls src_interface[:network_id] 3 times

File: vnet/lib/vnet/node_api/tunnel.rb [hunk 6]
```diff
@@ -15,17 +16,38 @@ def update_mode(id, mode)
       end
 
       #
+      # Override find method to only use the unique identifiers to find an existing tunnel
+      #
+      def find(options)
```
Comment (hansode) [inline vnet/lib/vnet/node_api/tunnel.rb]: [reek] [UtilityFunction] doesn't depend on instance state

File: vnet/lib/vnet/core/tunnel_manager.rb [hunk 7]
```diff
@@ -146,9 +146,9 @@ def item_initialize(item_map)
 
       item_class =
         case tunnel_mode
-        when :gre     then Tunnels::Gre
-        when :mac2mac then Tunnels::Mac2Mac
-        when :unknown then Tunnels::Unknown
+        when "gre"     then Tunnels::Gre
+        when "mac2mac" then Tunnels::Mac2Mac
+        when "unknown" then Tunnels::Unknown
```
Comment (Metallion) [inline vnet/lib/vnet/core/tunnel_manager.rb]: In this case, wouldn't it be better to check for both symbol and string?

```ruby
when :gre, "gre"     then Tunnels::Gre
```

File: vnet/lib/vnet/node_api/tunnel.rb [hunk 8]
```diff
@@ -15,21 +16,51 @@ def update_mode(id, mode)
...
+        tunnel = super({
+          :src_datapath_id => options[:src_datapath_id],
+          :dst_datapath_id => options[:dst_datapath_id],
+          :src_interface_id => options[:src_interface_id],
+          :dst_interface_id => options[:dst_interface_id]
+        })
+      end
+
+      def create_or_find(options)
+        # Only create a tunnel if it doesn't exist yet
```
Comment (hansode) [inline vnet/lib/vnet/node_api/tunnel.rb]: [rails_best_practices] remove trailing whitespace

File: vnet/lib/vnet/node_api/tunnel.rb [hunk 9]
```diff
@@ -15,21 +16,51 @@ def update_mode(id, mode)
...
       end
 
       def dispatch_deleted_item_events(model)
-        # dispatch_event(REMOVED_TUNNEL, id: model.id)
+         dispatch_event(REMOVED_TUNNEL, id: model.id)
 
         # no dependencies
       end
 
+      def sanitize_options(options)
```
Comment (hansode) [inline vnet/lib/vnet/node_api/tunnel.rb]: [reek] [FeatureEnvy] refers to options more than self

File: vnet/lib/vnet/node_api/tunnel.rb (hunk 3 above)
Comment (hansode) [inline vnet/lib/vnet/node_api/tunnel.rb]: [rails_best_practices] remove trailing whitespace

**Comment by Metallion** [comment]
I am still a little worried about changing symbols to strings like this and using a Mutex. From what I understand Celluloid should do all the thread locking for us. I am going to do some research about that before merging this.

**Comment by Mopster** [comment]
This PR became obsolete because of PR #305

**Pull request closed without merging**